sdk/resource/resource.go

```diff
@@ -43,20 +46,45 @@ var (
 	}(Detect(context.Background(), defaultServiceNameDetector{}, fromEnv{}, telemetrySDK{}))
 )
 
+var (
+	errMergeConflictSchemaURL = errors.New("cannot merge resource due to conflicting Schema URL")
+)
+
 // New returns a Resource combined from the user-provided detectors.
 func New(ctx context.Context, opts ...Option) (*Resource, error) {
 	cfg := config{}
 	for _, opt := range opts {
 		opt.apply(&cfg)
 	}
 
-	return Detect(ctx, cfg.detectors...)
+	resource, err := Detect(ctx, cfg.detectors...)
+
+	var err2 error
+	resource, err2 = Merge(resource, &Resource{schemaURL: cfg.schemaURL})
+	if err == nil {
+		err = err2
+	} else if err2 != nil {
+		err = fmt.Errorf("detecting resources: %s", []string{err.Error(), err2.Error()})
+	}
+
+	return resource, err
+}
+
+// NewWithAttributes creates a resource from attrs and associates the resource with a
+// schema URL. If attrs contains duplicate keys, the last value will be used. If attrs
+// contains any invalid items those items will be dropped. The attrs are assumed to be
+// in a schema identified by schemaURL.
+func NewWithAttributes(schemaURL string, attrs ...attribute.KeyValue) *Resource {
+	resource := NewSchemaless(attrs...)
+	resource.schemaURL = schemaURL
+	return resource
 }
 
-// NewWithAttributes creates a resource from attrs. If attrs contains
-// duplicate keys, the last value will be used. If attrs contains any invalid
-// items those items will be dropped.
-func NewWithAttributes(attrs ...attribute.KeyValue) *Resource {
+// NewSchemaless creates a resource from attrs. If attrs contains duplicate keys,
+// the last value will be used. If attrs contains any invalid items those items will
+// be dropped. The resource will not be associated with a schema URL. If the schema
+// of the attrs is known use NewWithAttributes instead.
+func NewSchemaless(attrs ...attribute.KeyValue) *Resource {
```

Comment:
nit: It seems a bit odd to name a function based on what it lacks (a schema).  It seems like `NewWithSchema` and `NewWithAttributes` would be easier to understand.

Reply:
Discussed earlier, see https://github.com/open-telemetry/opentelemetry-go/pull/1938#discussion_r637239381

The problem is that the newly introduced schema is both with schema and with attributes so we would have to call it NewWithSchemaAndAttributes which seems too long.

I am open to better suggestions, however neither NewWithAttributes nor NewWithSchema fully reflect the intent.

Slight possible benefit of NewSchemaless is that it kind of hopefully makes people pause and think about what they are doing. Why is it "schemaless"? Should I have a schema?

Reply:
> Slight possible benefit of NewSchemaless is that it kind of hopefully makes people pause and think about what they are doing: why is it "schemaless"? Should I have a schema?

⬆️⬆️⬆️⬆️